File: lib/storyarn_web/components/instruction_builder.ex
defmodule StoryarnWeb.Components.InstructionBuilder do
  @moduledoc """
  Thin HEEx wrapper that renders the InstructionBuilder JS hook container.

  The JS hook owns the assignments array client-side and pushes changes
  via `pushEvent("update_instruction_builder", %{assignments: [...]})`.

  Translations are passed via `data-translations` so the JS hook
  can render localized operator labels and UI strings.
  """

  use Phoenix.Component
  use Gettext, backend: StoryarnWeb.Gettext

  attr :id, :string, required: true
  attr :assignments, :list, default: []
  attr :variables, :list, default: []
  attr :can_edit, :boolean, default: true
  attr :context, :map, default: %{}
  attr :event_name, :string, default: nil

  def instruction_builder(assigns) do
    assigns = assign(assigns, :translations, translations())

    ~H"""
    <div
      id={@id}
      phx-hook="InstructionBuilder"
      phx-update="ignore"
      data-assignments={Jason.encode!(@assignments)}
      data-variables={Jason.encode!(@variables)}
      data-can-edit={Jason.encode!(@can_edit)}
      data-context={Jason.encode!(@context)}
      data-event-name={@event_name}
      data-translations={Jason.encode!(@translations)}
      class="instruction-builder"
    >
    </div>
    """
  end

  @doc false
  def translations do
    %{
      operator_verbs: %{
        "set" => dgettext("flows", "Set"),
        "add" => dgettext("flows", "Add"),
        "subtract" => dgettext("flows", "Subtract"),
        "set_true" => dgettext("flows", "Set"),
        "set_false" => dgettext("flows", "Set"),
        "toggle" => dgettext("flows", "Toggle"),
        "clear" => dgettext("flows", "Clear")
      },
      operator_dropdown_labels: %{
        "set" => dgettext("flows", "Set \u2026 to"),
        "add" => dgettext("flows", "Add \u2026 to"),
        "subtract" => dgettext("flows", "Subtract \u2026 from"),
        "set_true" => dgettext("flows", "Set \u2026 to true"),
        "set_false" => dgettext("flows", "Set \u2026 to false"),
        "toggle" => dgettext("flows", "Toggle"),
        "clear" => dgettext("flows", "Clear")
      },
      sentence_texts: %{
        "to" => dgettext("flows", "to"),
        "from" => dgettext("flows", "from"),
        "to true" => dgettext("flows", "to true"),
        "to false" => dgettext("flows", "to false")
      },
      add_assignment: dgettext("flows", "Add assignment"),
      no_assignments: dgettext("flows", "No assignments"),
      placeholder_sheet: dgettext("flows", "sheet"),
      placeholder_variable: dgettext("flows", "variable"),
      placeholder_value: dgettext("flows", "value"),
      switch_to_literal: dgettext("flows", "Switch to literal value"),
      switch_to_variable_ref: dgettext("flows", "Switch to variable reference")
    }
  end
end

File: assets/js/hooks/instruction_builder.js
/**
 * LiveView hook for the Instruction Builder.
 *
 * Thin wrapper around instruction_builder_core.js — delegates all rendering
 * logic to the core module while providing the LiveView hook interface.
 *
 * Communication:
 * - Receives initial data from data-* attributes (read in mounted())
 * - Pushes: "update_instruction_builder" event with { assignments: [...] }
 * - Listens: "node_updated" push event for collaboration updates
 *
 * The element uses phx-update="ignore" so LiveView won't clear
 * JS-rendered children on re-render.
 */

import { createInstructionBuilder } from "../screenplay/builders/instruction_builder_core.js";
import { deepEqual } from "../utils/deep_equal.js";

const DEFAULT_TRANSLATIONS = {
  add_assignment: "Add assignment",
  no_assignments: "No assignments",
  placeholder_sheet: "sheet",
  placeholder_variable: "variable",
  placeholder_value: "value",
};

export const InstructionBuilder = {
  mounted() {
    this.nodeId = null;
    this.assignments = JSON.parse(this.el.dataset.assignments || "[]");
    this.variables = JSON.parse(this.el.dataset.variables || "[]");
    this.canEdit = JSON.parse(this.el.dataset.canEdit || "true");
    this.context = JSON.parse(this.el.dataset.context || "{}");
    this.eventName = this.el.dataset.eventName || null;
    this._pendingPushCount = 0;

    this.t = {
      ...DEFAULT_TRANSLATIONS,
      ...JSON.parse(this.el.dataset.translations || "{}"),
    };

    // Extract node ID from element ID (format: "instruction-builder-{nodeId}")
    const idMatch = this.el.id.match(/instruction-builder-(\d+)/);
    if (idMatch) this.nodeId = parseInt(idMatch[1], 10);

    // Resolve the event name
    const resolvedEventName = this.eventName || "update_instruction_builder";

    const pushEvent = (name, payload) => {
      this._pendingPushCount++;

      if (this.eventName) {
        this.pushEvent(name, payload);
      } else {
        this.pushEvent(resolvedEventName, {
          assignments: payload.assignments,
        });
      }
    };

    this.builderInstance = createInstructionBuilder({
      container: this.el,
      assignments: this.assignments,
      variables: this.variables,
      canEdit: this.canEdit,
      context: this.context,
      eventName: resolvedEventName,
      pushEvent,
      translations: this.t,
    });

    // Listen for server push events (collaboration updates)
    this.handleEvent("node_updated", (data) => {
      if (data.id !== this.nodeId) return;

      if (this._pendingPushCount > 0) {
        this._pendingPushCount--;
        return;
      }

      const newAssignments = data.data?.assignments || [];
      if (!deepEqual(newAssignments, this.assignments)) {
        this.assignments = newAssignments;
        this.builderInstance.update(newAssignments);
      }
    });
  },

  destroyed() {
    if (this.builderInstance) {
      this.builderInstance.destroy();
      this.builderInstance = null;
    }
  },
};
